File: 20-valid-parentheses/valid-parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        Dictionary<char, char> closed_to_open = new Dictionary<char, char>();
        closed_to_open.Add(')', '(');
        closed_to_open.Add('}', '{');
        closed_to_open.Add(']', '[');

        Stack<char> stack = new Stack<char>();

        for (int i = 0; i < s.Length; i++)
        {
            char parenthese = s[i];

            if (closed_to_open.ContainsKey(parenthese))
            {
                // Closed parenthese
                if (stack.Count == 0)
                    return false;
                
                char innermost_parenthese = stack.Pop();
                if (closed_to_open[parenthese] != innermost_parenthese)
                    return false;
            }
            else
            {
                // Open parenthese
                stack.Push(parenthese);
            }
        }

        return stack.Count == 0;
    }
}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        closed_to_open = {
            ")": "(",
            "}": "{",
            "]": "["
        }
        
        for parenthese in s:
            if parenthese in "({[":
                stack.append(parenthese)
            else:
                if len(stack) == 0 or stack[-1] != closed_to_open[parenthese]:
                    return False
                stack.pop()

        return len(stack) == 0